src/embeddings/tools/sgns.cpp

```diff
@@ -0,0 +1,573 @@
+typedef std::deque<sgns_vocab_vector::size_type> sgns_window;
+
+// Represents the neural network that is trained.
+struct sgns_net
+{
+    sgns_net(std::size_t vocab_size, std::size_t layer1_size) :
+        syn0(vocab_size * layer1_size),
+        syn1neg(vocab_size * layer1_size, 0)
+    {
+        std::default_random_engine random; // TODO: Seed
+        std::uniform_real_distribution<float> uniform(-0.5f / (float)layer1_size,
+                                                       0.5f / (float)layer1_size);
+
+        for (auto& i : syn0)
+            i = uniform(random);
+    }
+
+    sgns_net_vector syn0;
+    sgns_net_vector syn1neg;
+};
+
+// The primary class for executing SGNS.
+class sgns_trainer
+{
+  public:
+
+    // Constructor.
+    sgns_trainer(const cpptoml::table& cfg,
+                 const cpptoml::table& embed_cfg,
+                 const cpptoml::table& sgns_cfg) :
+        cfg_(cfg),
+        embed_cfg_(embed_cfg),
+        prefix_(*embed_cfg.get_as<std::string>("prefix")),
+        vector_size_(embed_cfg.get_as<std::size_t>("vector-size").value_or(100)),
+        num_threads_(embed_cfg.get_as<std::size_t>("num-threads").value_or(std::max(1u, std::thread::hardware_concurrency()))),
+        max_ram_(embed_cfg.get_as<std::size_t>("max-ram").value_or(4096) * 1024 * 1024),
+        subsample_threshold_(sgns_cfg.get_as<double>("subsample-threshold").value_or(1E-4)),
+        max_window_size_(sgns_cfg.get_as<std::size_t>("max-window-size").value_or(6)),
+        iterations_(sgns_cfg.get_as<uint64_t>("iterations").value_or(10)),
+        starting_learning_rate_((float)sgns_cfg.get_as<double>("learning-rate").value_or(0.025)),
+        negative_samples_(sgns_cfg.get_as<std::size_t>("negative-samples").value_or(20)),
+        binary_output_(sgns_cfg.get_as<bool>("binary").value_or(true)),
+        vocab_(load_vocab(prefix_ + "/vocab.bin")),
+        noise_dist_(create_unigram_noise_distribution(1E8)), // TODO: Configurable size?
+        net_(vocab_.vector.size(), vector_size_),
+        learning_rate_(starting_learning_rate_),
+        word_count_actual_(0)
+    {
+        std::mutex io_mutex;
+        printing::progress progress{" > Training: ", (uint64_t)(starting_learning_rate_ * 100000.0)};
+
+        train(io_mutex, progress);
+
+        save_w2v_vectors();
+        save_meta_vectors();
+    }
+
+    // When the thread_pool goes out of scope, the threads will join. In other
+    // words, when this method returns, all threads have finished training.
+    void train(std::mutex& io_mutex, printing::progress& progress)
+    {
+        auto stream = analyzers::load_filters(cfg_, embed_cfg_);
+        if (!stream)
+        {
+            throw sgns_exception{"Failed to find an ngram-word analyzer configuration"};
+        }
+
+        parallel::thread_pool pool{num_threads_};
+
+        auto iterations_left = iterations_;
+
+        auto docs = corpus::make_corpus(cfg_);
```

Comment:
Can we actually remove the `reset()` and `reset_metadata()` methods unless my profiling indicates they are actually needed as a performance improvement? They are only needed here, and many of the corpora don't support them, so it's confusing from a use perspective. A common use case is likely going to be to run embeddings on compressed corpora, since you want to have a lot of raw text to train with.

(It may end up being the case that we don't even operate on a corpus here eventually, as I suspect the repeated tokenization process may be what slows us down relative to the original word2vec.)